ddtrace/internal/forksafe.py

```diff
@@ -0,0 +1,106 @@
+import inspect
+import functools
+import os
+import threading
+
+
+__all__ = [
+    "call_nocheck",
+    "ddtrace_after_in_child",
+    "register",
+]
+
+
+BUILTIN_FORK_HOOKS = hasattr(os, "register_at_fork")
+
+if BUILTIN_FORK_HOOKS:
+    registry = []
+
+    def ddtrace_after_in_child():
+        global registry
+
+        for fn, hook in registry:
+            hook()
+            fn._fork_occurred = True
+
+    def register(after_in_child):
+        # TODO: this function won't work if not used as a decorator.
+        # eg: register(hook) won't result in hook being called.
+        def wrapper(func):
+            args = inspect.getfullargspec(func)[0]
+            include_arg = "is_in_child_after_fork" in args
+
+            @functools.wraps(func)
+            def fsfunc(*args, **kwargs):
+                fork_occurred = fsfunc._fork_occurred
+                fsfunc._fork_occurred = False
+                if include_arg:
+                    return func(*args, is_in_child_after_fork=fork_occurred, **kwargs)
+                else:
+                    return func(*args, **kwargs)
+
+            fsfunc._fork_occurred = False
+            registry.append((fsfunc, after_in_child))
+            return fsfunc
+
+        return wrapper
+
+    def call_nocheck(f, *args, **kwargs):
+        return f(*args, **kwargs)
```

Comment:
I missed the point of this function?

Reply:
It's mainly for the API when built-in hooks are not available. It allows the function to be called directly (by-passing the PID check) if the user knows what they're doing. In the case of there being built-in hooks obviously no check has to be done, so it's a no-op.

Reply:
Oh, though I don't see that being used anywhere? Should we really provide it then?

Reply:
I think that still stands :)

Reply:
I think it's been covered in the other comments now 🙂

Reply:
Sorry, I still don't see where/why this is going to be used?

Reply:
It by-passes the PID check in case someone knows that the fork checks will be run already or they don't care. A concrete example of where this will be used is with `_rand.rand64bits()`. In the tracer we call `reseed` for the RNG in `Tracer.start_span` and then downstream in `Span.__init__` we want to optionally skip doing the check since we know a fork occurred already as we pass a flag down to `Span.__init__`.

Reply:
Isn't that premature optimization and complicating the workflow for little gain?